tests/functional/test_s3.py

```diff
@@ -225,6 +244,227 @@ def progress_callback(amount):
+        # contents needs to be pruned, and Range needs to be an expected param.
+        if end_byte is not None:
+            contents = full_contents[start_byte:end_byte+1]
+            part_range = 'bytes=%s-%s' % (start_byte, end_byte_range)
+            content_range = 'bytes=%s-%s/%s' % (
+                start_byte, end_byte, len(full_contents))
+            get_object_response['ContentRange'] = content_range
+            expected_params['Range'] = part_range
+
+        get_object_response.update({
+            "AcceptRanges": "bytes",
+            "ETag": self.etag,
+            "ContentLength": len(contents),
+            "ContentType": "binary/octet-stream",
+            "Body": six.BytesIO(contents),
+            "ResponseMetadata": {
+                "HTTPStatusCode": 200
+            }
+        })
+        expected_params.update({
+            "Bucket": self.bucket,
+            "Key": self.key
+        })
+
+        self.stubber.add_response(
+            method='get_object', service_response=get_object_response,
+            expected_params=expected_params)
+
+    def stub_multipart_download(self, contents, part_size, num_parts):
+        self.stub_head(content_length=len(contents))
+
+        for i in range(num_parts):
+            start_byte = i * part_size
+            end_byte = (i + 1) * part_size - 1
+            self.stub_get_object(
+                full_contents=contents, start_byte=start_byte,
+                end_byte=end_byte)
+
+    def test_client_download(self):
+        self.stub_single_part_download()
+        with self.stubber:
+            self.s3.meta.client.download_fileobj(
+                Bucket=self.bucket, Key=self.key, Fileobj=self.fileobj)
+
+        self.assertEqual(self.fileobj.getvalue(), self.contents)
+
+    def test_bucket_download(self):
+        self.stub_single_part_download()
+        bucket = self.s3.Bucket(self.bucket)
+        with self.stubber:
+            bucket.download_fileobj(Key=self.key, Fileobj=self.fileobj)
+
+        self.assertEqual(self.fileobj.getvalue(), self.contents)
+
+    def test_object_download(self):
+        self.stub_single_part_download()
+        obj = self.s3.Object(self.bucket, self.key)
+        with self.stubber:
+            obj.download_fileobj(Fileobj=self.fileobj)
+
+        self.assertEqual(self.fileobj.getvalue(), self.contents)
+
+    def test_multipart_download(self):
+        self.contents = b'A' * 55
+        self.stub_multipart_download(
+            contents=self.contents, part_size=5, num_parts=11)
+        transfer_config = TransferConfig(
+            multipart_chunksize=5, multipart_threshold=1,
+            max_concurrency=1)
+
+        with self.stubber:
+            self.s3.meta.client.download_fileobj(
+                Bucket=self.bucket, Key=self.key, Fileobj=self.fileobj,
+                Config=transfer_config)
+
+        self.assertEqual(self.fileobj.getvalue(), self.contents)
+
+    def test_download_progress(self):
```

Comment:
It may be good to have a progress test for uploads as well since it does not share the same code path as ``S3Transfer``. To simulate the reading of the uploaded body is a little tricky, you could probably do something like I did for the s3transfer tests: https://github.com/boto/s3transfer/blob/develop/tests/functional/test_upload.py#L61